### PR TITLE
Add core framework for Node Health Checker (NHC)

### PR DESCRIPTION
**Pull request by banool**
## Context
This PR aims to implement part of the [feature set necessary for the node health checker (NHC) to support AIT2](https://www.notion.so/aptoslabs/Target-for-AIT2-d2e22a218c074e84a2be9a882ceb07f2). Notably missing right now are most of the metrics checks, so far I've focused on making a broad framework that can handle the wide range of different use cases we need to support. I recommend you check out that doc before tackling this PR.

## Description
The stack in this PR up to this point comprises an instance of the node health checker (NHC) the following features:
- A trait defining raw metric collection.
- An implementation of that trait that can collect metrics from nodes with reqwest.
- A trait defining metric evaluation, in which each evaluator takes in two rounds of metrics from the baseline and target nodes and returns a set of evaluation results.
- An implementation of that trait for checking state sync metrics.
- A trait defining a runner that coordinates metric collection and evaluation.
- A basic implementation of that runner that collects metrics, waits, and collects metrics again, parses them with prometheus_parse, passes them to the evaluators, and finally returns an overall evaluation.
- Full configurability for every component so far via the CLI, such as the metric collection delay, state sync version lag threshold, etc.
- The ability to run NHC as a sidecar, where you configure it from startup with a specific target node. You can lock the instance down such that it only ever evaluates this node.
- The ability to select which evaluators to use on a per-instance basis. This is necessary since some evaluators might not be relevant for certain node types (e.g. consensus isn't relevant for full nodes).
- OpenAPI support. This is how we do things with the regular API, and from speaking with Max it seemed like an easy approach for NHC as well. This was the main reason I decided to go with Poem (which is otherwise kind of new). Saying that, the framework has been fantastic so far.

## Test Plan
Check out the README in this stack, it details how I've been testing things so far.

## PR disclaimer
GitHub doesn't really handle stacks super well. My choices were either a single PR with a bunch of commits or a bunch of PRs with a single PR. I would prefer the latter, since I can have a summary and test plan for each change, but git makes it pretty hard, so hopefully this is okay. I've tried to make a comprehensive summary to capture all the changes.

I've got a pretty good handle of stacking commits in git now lol so I encourage you to review the stack on a commit-by-commit basis, but also feel free to just leave a broad review of the sum changes. I'll try to make any further changes (except those directly addressing this PR) in another branch based off this one.

### CODE REVIEW

**Comment by banool** [comment]
I've decided to build NHC such that we don't need multiple deployments, we just have one NHC instance that can handle all the different node / network types. See NHC instance unification in https://www.notion.so/aptoslabs/Target-for-AIT2-d2e22a218c074e84a2be9a882ceb07f2. I can share more info here for external contributors if they're interested too.

File: ecosystem/node-checker/src/metric_collector/common.rs
Comment (JoshLind) [inline ecosystem/node-checker/src/metric_collector/common.rs:8]: Where is this used? Grep suggests it's not 🤔

File: ecosystem/node-checker/src/main.rs
Comment (JoshLind) [inline ecosystem/node-checker/src/main.rs:62]: Would we ever be able to support `https:`?

File: ecosystem/node-checker/src/args.rs
Comment (JoshLind) [inline ecosystem/node-checker/src/args.rs:53]: - `only the todo endpoint will work`? What does this mean?
- Also `allow_test_node_only`? This should be, `allow_preconfigured_test_node_only`, right?

File: ecosystem/node-checker/src/main.rs
Comment (JoshLind) [inline ecosystem/node-checker/src/main.rs:38]: nit: can we replace this with a TODO? Just in case someone starts pinging/running it?

File: ecosystem/node-checker/src/main.rs
Comment (JoshLind) [inline ecosystem/node-checker/src/main.rs:53]: Can we rename this? I find the term `test` confusing. Why not just `allow_preconfigured_node_only`?

File: ecosystem/node-checker/README.md
Comment (JoshLind) [inline ecosystem/node-checker/README.md:7]: FYI: I think the README should begin with instructions on how to use the tool. As opposed to information about development. But, we can do this later.

File: ecosystem/node-checker/src/metric_evaluator/traits.rs
Comment (JoshLind) [inline ecosystem/node-checker/src/metric_evaluator/traits.rs:25]: Will deriving `Display` not do the trick?

File: ecosystem/node-checker/src/metric_evaluator/traits.rs
Comment (JoshLind) [inline ecosystem/node-checker/src/metric_evaluator/traits.rs:31]: thisError :P

**Review by JoshLind — CHANGES_REQUESTED** [review]
Awesome, thanks @banool! 🚀
- Most of my comments are around readability/nits. Nothing super worrying.
- Also, for future reference, we generally try to keep PRs under 1k lines of code (unless the code is generated). I know we're trying to move fast, but it reduces reviewer burden 😄

File: ecosystem/node-checker/DEV_NOTES.md
Comment (banool) [reply ecosystem/node-checker/DEV_NOTES.md:1]: @JoshLind Not sure if you saw my PR disclaimer, I tried to address this hahah. I wrote this stack with the intent that each commit would be reviewed individually. I could make 20ish PRs but this feels like a lot of overhead. I also plan to generally keep my commit history clean such that I rebase, not squash. I was chatting with David recently and he was team rebase also (if your stack is clean enough, which this may not be). Perhaps we need to discuss as a team which approach to go with, I'm used to stack based workflows where every commit should pass tests, have a commit message, test plan, etc.

File: ecosystem/node-checker/src/main.rs
Comment (banool) [reply ecosystem/node-checker/src/main.rs:53]: Test is a little confusing agreed, but I want to qualify it with something to make it clear that this is the node being tested. I've toyed with "investigation" and "target" but they don't seem much better.

File: ecosystem/node-checker/src/metric_evaluator/state_sync_evaluator.rs
Comment (banool) [reply ecosystem/node-checker/src/metric_evaluator/state_sync_evaluator.rs:95]: This is just gross nesting, that first closing brace is for `Evaluation`. I'll fix that up.

File: ecosystem/node-checker/src/metric_evaluator/traits.rs
Comment (banool) [reply ecosystem/node-checker/src/metric_evaluator/traits.rs:25]: Display cannot be derived sadly.

File: ecosystem/node-checker/src/metric_evaluator/traits.rs
Comment (banool) [reply ecosystem/node-checker/src/metric_evaluator/traits.rs:31]: Yeah it's on the todo list, it just felt sorta heavy at the time. Mixing thiserror and anyhow error feels sorta bad.

File: ecosystem/node-checker/src/public_types.rs
Comment (banool) [reply ecosystem/node-checker/src/public_types.rs:8]: This is used when generating the OpenAPI spec, it will go to the Example impl for this struct and use that to build the example string.

File: ecosystem/node-checker/src/public_types.rs
Comment (banool) [reply ecosystem/node-checker/src/public_types.rs:69]: Complete in this sense is meant to be like "TotalEvaluation" or "EvaluationSummary". I can try think of a better name.

File: ecosystem/node-checker/src/runner/blocking_runner.rs
Comment (banool) [reply ecosystem/node-checker/src/runner/blocking_runner.rs:97]: Not to get too philosophical but I wouldn't say rustiness is defined by using a lot of combinators hahaha, I'm a firm believer in the readability of the for loop, I only use `map` if it can fit on one line.

File: ecosystem/node-checker/DEV_NOTES.md
Comment (geekflyer) [reply ecosystem/node-checker/DEV_NOTES.md:1]: I think 5 of your commits are something like `run cargo fmt` , `address lint`, `touch up X` , so those are not the kind of commits I'd like to see in main :)
In the current state I'd definitely squash this PR, but if you can reduce this to a few fewer commits it's fine rebase this instead.

One thing I wanted to call out the GitHub - for better or worse - is very PR oriented and the PR is the unit of change and contains most of the context.
I know at Meta you used phab and stacked diffs but GitHub just isn't really built with these workflows (such as stacking PRs) in mind and trying to bend GitHub to work like phabricator's review flow isn't trivial.

Actually on that note, if we were to get rid of bors/the merge queue, we could use something like this https://github.com/ejoffe/spr for people who like stacking PRs (I tried this out and it's pretty cool but it doesn't work with a merge queue). cc @davidiw  fyi

File: ecosystem/node-checker/README.md
Comment (banool) [reply ecosystem/node-checker/README.md:7]: Agreed, but for that I'd want to have terraform configs built out first.

File: ecosystem/node-checker/src/args.rs
Comment (banool) [reply ecosystem/node-checker/src/args.rs:24]: Normally I have this if the tool is something you run once, like a CLI, and otherwise not. While at this commit this falls into the latter bucket, in later commits I add other subcommands that you would run once (e.g. `nhc configuration create`), in which case I think `-d` is a nice helper vs having to set a couple of env vars with every CLI invocation.

File: ecosystem/node-checker/src/args.rs
Comment (banool) [reply ecosystem/node-checker/src/args.rs:34]: Yeah I had the same idea, I do that in https://github.com/aptos-labs/aptos-core/pull/1418.

File: ecosystem/node-checker/DEV_NOTES.md
Comment (banool) [reply ecosystem/node-checker/DEV_NOTES.md:1]: Yeah this stack isn't really in a rebaseable state right now, agreed. I suppose committing to squashing does make my life easier because I don't have to make a clean stack, just a clean PR once it's all merged.

File: ecosystem/node-checker/src/main.rs
Comment (banool) [reply ecosystem/node-checker/src/main.rs:62]: Yeah especially if the operator has considered some kind of https proxy in front. This actually becomes redundant in a later diff, I switch to using `Url` which requires a scheme, so the user can set this themselves.

File: ecosystem/node-checker/src/metric_evaluator/state_sync_evaluator.rs
Comment (banool) [reply ecosystem/node-checker/src/metric_evaluator/state_sync_evaluator.rs:13]: I'll leave a todo 👍

File: ecosystem/node-checker/src/metric_evaluator/traits.rs
Comment (banool) [reply ecosystem/node-checker/src/metric_evaluator/traits.rs]: Though thiserror impls it, which I just added 😄

File: ecosystem/node-checker/src/runner/blocking_runner.rs
Comment (JoshLind) [reply ecosystem/node-checker/src/runner/blocking_runner.rs:98]: > I only use map if it can fit on one line.

Your multiline use of map above contradicts this statement 😄 But, sure, it's not a biggie!

File: ecosystem/node-checker/src/main.rs
Comment (JoshLind) [reply ecosystem/node-checker/src/main.rs:53]: Yeah, the other options aren't great. I still think `allow_preconfigured_node_only` is clearer than `allow_preconfigured_test_node_only`, but I'll leave it up to you.

File: ecosystem/node-checker/src/args.rs
Comment (JoshLind) [reply ecosystem/node-checker/src/args.rs:24]: I guess the question in my mind is "Why would the user want debug logs?" to which the only answer I can come up with is "for debugging/development"?. If you suspect regular users (i.e., those who aren't developers) want to see debug logs, it feels to me like we should just log relevant information at a higher level as opposed to adding a flag that explicitly sets debug logs. One could wonder, why not a flag for all debug levels?

But, this isn't a blocker 😄

**Review by JoshLind — APPROVED** [review]
Thanks @banool. Looks good. I'll unblock this for now so we can continue to iterate! Please be sure to squash all of these commits. We don't want a 31 commit stream in the history :O

File: ecosystem/node-checker/DEV_NOTES.md
Comment (JoshLind) [reply ecosystem/node-checker/DEV_NOTES.md]: > I wrote this stack with the intent that each commit would be reviewed individually

In practice I find it much quicker to review the entire stack at once (might be me though). Often times PR authors (myself included) will do some work in a commit, then undo parts of the very same work in the next commit with a slight modification, etc. This tends to lead unnecessary work (especially for super large PRs like this).

File: ecosystem/node-checker/src/args.rs
Comment (banool) [reply ecosystem/node-checker/src/args.rs:24]: Ok yeah I think from that perspective it makes sense, I'll remove the flag 👍

File: ecosystem/node-checker/DEV_NOTES.md
Comment (banool) [reply ecosystem/node-checker/DEV_NOTES.md]: Yeah sounds good, I realise with GitHub it's easier to treat one PR as one unit of review. My upcoming PRs are much smaller accordingly.

**Comment by banool** [comment]
aiya

**Comment by banool** [comment]
It's been 4 days so maybe it'll magically land now, trying one more time. Then I'll debut properly. /land

**Comment by banool** [comment]
I got it this time, it was a dockerignore issue.

**Comment by banool** [comment]
image

we made it

**Pull request closed without merging**
